File: 14-TypeScript/Grey Multi-Tenant/packages/grey-swift/Sources/GreySDK/GreySDK.swift
// =============================================================================
// Grey Multi-Tenant SDK - Swift
// GreySDK.swift (Entrypoint)
//
// Main entrypoint and public API exports for the Grey SDK.
// =============================================================================

import Foundation

// =============================================================================
// Public API Exports
// =============================================================================

// This file serves as the main module entrypoint.
// All public types are exported from their respective files.

// MARK: - Configuration
// - GreyOptions
// - GreyOptions.Builder

// MARK: - Error Handling
// - GreyError
// - GreyResult

// MARK: - Main Client
// - GreyClient

// MARK: - Domain Clients
// - AuthClient
// - UserClient
// - ProjectsClient
// - QueryClient
// - MutationClient

// MARK: - Domain Models
// - AuthData
// - User
// - Project
// - Pagination
// - ProjectsData
// - QueryData
// - MutationData

// =============================================================================
// SDK Info
// =============================================================================

/// Grey Multi-Tenant SDK information.
public enum GreySDK {
    
    /// SDK version.
    public static let version = "0.1.0"
    
    /// SDK name.
    public static let name = "GreySDK"
    
    /// Create a new Grey client.
    /// - Parameters:
    ///   - host: The gRPC server host
    ///   - port: The gRPC server port (default: 443)
    ///   - useTLS: Whether to use TLS (default: true)
    /// - Returns: GreyClient instance
    /// - Throws: If channel creation fails
    public static func client(
        host: String,
        port: Int = 443,
        useTLS: Bool = true
    ) throws -> GreyClient {
        try GreyClient(options: GreyOptions(
            host: host,
            port: port,
            useTLS: useTLS
        ))
    }
    
    /// Create a client for local development.
    /// - Parameter port: The local server port (default: 50051)
    /// - Returns: GreyClient configured for local development
    /// - Throws: If channel creation fails
    public static func localClient(port: Int = 50051) throws -> GreyClient {
        try GreyClient.forLocalDev(port: port)
    }
}

// =============================================================================
// Usage Examples (Documentation)
// =============================================================================

/*
 Quick Start:
 
 ```swift
 import GreySDK
 
 // Create client
 let grey = try GreySDK.client(host: "api.grey.example.com")
 
 // Login
 do {
     let auth = try await grey.auth.login(email: "user@example.com", password: "password")
     print("Logged in as: \(auth.userId)")
 } catch let error as GreyError {
     print("Login failed: [\(error.code)] \(error.message)")
 }
 
 // Fetch user
 let user = try await grey.user.getUser()
 print("User: \(user.name ?? "Unknown")")
 
 // List projects
 let projectsData = try await grey.projects.listProjects()
 for project in projectsData.projects {
     print("Project: \(project.name)")
 }
 
 // Create project
 let newProject = try await grey.projects.createProject(
     name: "My Project",
     description: "A new project"
 )
 
 // Generic query
 let queryResult = try await grey.query.query(
     endpoint: "/custom/endpoint",
     params: ["key": "value"]
 )
 
 // Generic mutation
 let mutationResult = try await grey.mutation.post(
     endpoint: "/custom/action",
     body: ["data": "value"]
 )
 
 // Cleanup
 try await grey.shutdown()
 ```
 
 Builder Pattern:
 
 ```swift
 let grey = try GreyClient.create { builder in
     builder.host("api.grey.example.com")
            .port(443)
            .useTLS(true)
            .timeoutSeconds(30)
 }
 ```
 
 Local Development:
 
 ```swift
 let grey = try GreySDK.localClient(port: 50051)
 ```
 
 Session Restoration:
 
 ```swift
 // Load tokens from Keychain
 if let accessToken = loadAccessToken(),
    let refreshToken = loadRefreshToken() {
     grey.restoreAuth(accessToken: accessToken, refreshToken: refreshToken)
 }
 ```
 */

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-dart/lib/grey_sdk.dart
/// Grey Multi-Tenant SDK for Dart.
///
/// A Dart SDK for the Grey Multi-Tenant platform using gRPC transport.
///
/// ## Getting Started
///
/// ```dart
/// import 'package:grey_sdk/grey_sdk.dart';
///
/// void main() async {
///   final client = GreyClient(
///     GreyOptions(host: 'api.grey.example.com'),
///   );
///
///   try {
///     final auth = await client.auth.login(
///       email: 'user@example.com',
///       password: 'password',
///     );
///     print('Logged in: ${auth.accessToken}');
///
///     final user = await client.user.getUser();
///     print('User: ${user.name}');
///   } on GreyError catch (e) {
///     print('Error: [${e.code}] ${e.message}');
///   } finally {
///     await client.shutdown();
///   }
/// }
/// ```
library grey_sdk;

// Config
export 'src/config/grey_options.dart';

// Error handling
export 'src/error/grey_error.dart';

// Types
export 'src/grpc/types.dart';

// Main client
export 'src/grey_client.dart';

// Domain clients (for type access)
export 'src/domain/auth_client.dart';
export 'src/domain/mutation_client.dart';
export 'src/domain/projects_client.dart';
export 'src/domain/query_client.dart';
export 'src/domain/user_client.dart';
